### PR TITLE
Tab group fixes

### PR DESCRIPTION
**Pull request by rddunphy**
I found some time to tackle some of the low-hanging fruit from #4314:

- Add completion filtering to tab group commands
- Add an optional parameter to `:tgroupclose` to allow closing tab groups by name
- Make `:tgroupmove` create a new tab group rather than display an error if the specified tab group doesn't exist
- Display the current tab group (marked with suffix `(active)`) in the tab group completion list - this is consistent with the behaviour of `:tab` and other similar commands. 
- Clarify the behaviour of `:tgroupclose` in the documentation
- Switching to the current tab group now just does nothing, rather than displaying an error message - this is a bit less annoying if you have key bindings for tab groups.

I haven't figured out how to make the tab group in the status indicator update, which currently doesn't happen when you open a new tab or if you pin or unpin a tab - I'll have a go at that later this week, hopefully.

### CODE REVIEW

**Comment by rddunphy** [comment]
> * Display the current tab group (marked with suffix `(active)`) in the tab group completion list

Alternatively, we could use the prefix divs to mark the current and last tab groups with '%' and '#'. Then we could allow calling e.g. `tgroupmove #` to move a tab to the alternate tab group.

**Comment by bovine3dom** [comment]
> Alternatively, we could use the prefix divs to mark the current and last tab groups with '%' and '#'. Then we could allow calling e.g. tgroupmove # to move a tab to the alternate tab group.

Yeah, I think this makes more sense. It would be good if we could unify it as much as possible with `tab` completions - e.g. it would be nice to see which group was playing audio.

Otherwise this looks good to me - I'll merge it once the tests pass :)

**Comment by rddunphy** [comment]
>It would be good if we could unify it as much as possible with `tab` completions - e.g. it would be nice to see which group was playing audio.

Good idea - if you want to hold off on merging for now, I can have a go at this tomorrow.

**Comment by rddunphy** [comment]
I added an audio marker, and while I was there added the content div, which shows the url of the last active tab in each group. I think the other markers for `tab` probably don't make much sense for groups (e.g. marking a group as muted when probably only some tabs in that group are muted would just be confusing). 

Completions look like this now (with `completions.Tabs.statustylepretty=true`):

Screenshot_20230117_150020

**Comment by bovine3dom** [comment]
I'm thinking maybe we should include all URLs in the tab groups and sort them by last used, joining them with e.g. a comma. Happy to add it if you're too busy

**Comment by rddunphy** [comment]
Like this?

Screenshot_20230118_155635

**Comment by bovine3dom** [comment]
Yeah!  I think we should let people match by URLs too, probably?

I'm wondering if maybe there should be command like `:taball` that searches through all groups and then activates the selected tab along with its group ... but that should probably be a separate issue.

**Comment by rddunphy** [comment]
>Yeah! I think we should let people match by URLs too, probably?

That would be good - I'll take a look at the completion source for `:tab` and see if I can figure it out, but I'm not sure I'm fully understanding how the filtering is supposed to work. Are all completion sources that extend `Completions.CompletionSourceFuse` meant to do fuzzy matching? As far as I can tell, completion sources like `:help` only filter for exact matches using `.startsWith`: 
https://github.com/tridactyl/tridactyl/blob/2fd7f87c1ec9b320dbd96d1a2b439a27da810215/src/completions/Help.ts#L64

>I'm wondering if maybe there should be command like `:taball` that searches through all groups and then activates the selected tab along with its group ... but that should probably be a separate issue.

I think `:taball` already does that.

**Comment by bovine3dom** [comment]
> I think :taball already does that.

Oh, you're right. We should probably make it sort/group tabs by group to make it more obvious...

> Are all completion sources that extend Completions.CompletionSourceFuse meant to do fuzzy matching?

I _think_ only those with `sortScoredOptions = true` have fuzzy matching. Even then, fuzzy matching is only used for sorting the results. The filtering happens first and determines what is given to the fuzzy matcher.

I really dislike our completion code. I think the whole object-oriented nature of it makes it needlessly confusing, hard to configure it at runtime and encourages copy-pasted code (just look at Help.ts vs Apropos.ts ...) - and I one day would like to rewrite it. I started here but lost track of it: https://github.com/tridactyl/tridactyl/pull/3895/files

I'm sorry that you're having to try to understand it!

**Comment by rddunphy** [comment]
> We should probably make it sort/group tabs by group to make it more obvious...

We could maybe add markup for the tab group in a separate column? (Btw, is it intentional that the prefix column isn't populated for `:taball`?)

There's also the `tabshowhidden` setting, which will make all tabs appear in `:tab`. This seems to also cause a weird inconsistency I hadn't noticed before - with `tabshowhidden` false (the default), `:tab #` switches to the last active tab in the current group (because other groups aren't included in the completion), but `:bind t tab #` followed by `t` switches to the last active tab in any group. I'm not sure which is the "correct" behaviour. 

>I'm sorry that you're having to try to understand it!

No problem, just wanted to make sure I wasn't implementing the wrong class!

**Review by bovine3dom — CHANGES_REQUESTED** [review]
Looks great, just found one little bug :)

File: pretty.GeWrIIb3A/pretty.iHne78Dzi
Comment (rddunphy) [reply pretty.GeWrIIb3A/pretty.iHne78Dzi]: Oops

File: src/lib/tab_groups.ts
Comment (rddunphy) [reply src/lib/tab_groups.ts:88]: I think it's always safe to return undefined here - at least, I don't get that particular error any more.

**Comment by bovine3dom** [comment]
Fixed a bug where in `:taball` only the first 2-3 characters from the group name would appear in completions.

LGTM - will merge if/when tests pass. Thanks so much for all your work!

**Comment by rddunphy** [comment]
>Fixed a bug where in `:taball` only the first 2-3 characters from the group name would appear in completions.

This doesn't work if some tabs are in groups and others aren't (e.g. because they're pinned):
image

I can't see an easy way of hiding the entire column, unfortunately.

**Comment by bovine3dom** [comment]
If we put the tab groups at the end the problem goes away, are you OK with that?

**Comment by rddunphy** [comment]
Looks good to me! I've just made the column a bit narrower, as I don't think tab group names will usually to be that long. It still annoys me that I can't find a good way to make the column automatically fit the content. All the recommended solutions (e.g. [here](https://stackoverflow.com/questions/11267154/fit-cell-width-to-content)) don't seem to work for me.

**Comment by rddunphy** [comment]
Excellent, thank you!

**Pull request merged**
